src/artifacts.js

```diff
@@ -423,6 +643,119 @@ var replyWithArtifact = async function(taskId, runId, name, req, res) {
   this.monitor.reportError(err);
 };
 
+/** Complete artifact */
+api.declare({
+  method:     'put',
+  route:      '/task/:taskId/runs/:runId/artifacts/:name(*)',
+  name:       'completeArtifact',
+  stability:  API.stability.experimental,
+  scopes: [
+    [
+      'queue:create-artifact:<name>',
+      'assume:worker-id:<workerGroup>/<workerId>',
+    ], [
+      'queue:create-artifact:<taskId>/<runId>',
+    ],
+  ],
+  deferAuth:  true,
+  input:      'put-artifact-request.json#',
+  output:     'put-artifact-response.json#',
+  title:      'Complete Artifact',
+  description: 'tbd',
+}, async function(req, res) {
+  let taskId      = req.params.taskId;
+  let runId       = parseInt(req.params.runId, 10);
+  let name        = req.params.name;
+  let input       = req.body;
+  let isPublic    = /^public\//.test(name);
+
+  // Hmm, list destructuring is probably better here
+  let [artifact, task] = await Promise.all([
+    this.Artifact.load({taskId, runId, name}, true),
+    this.Task.load({taskId}, true),
+  ]);
+
+  // TODO: Have a shortcut here where things which are already
```

Comment:
That's just adding `if (artifact.present) { return res.status(200).json({});}`, maybe add it after checking that the run exists :)

Reply:
the number of times we'll hit this is very small, so let's leave the full codepath in place.